subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/PropertyValidationAccess.java

```diff
@@ -54,23 +55,34 @@
     );
 
     @SuppressWarnings("unused")
-    public static void collectTaskValidationProblems(Class<?> beanClass, Map<String, Boolean> problems) {
+    public static void collectTaskValidationProblems(Class<?> topLevelBean, Map<String, Boolean> problems) {
         DefaultTaskClassInfoStore taskClassInfoStore = new DefaultTaskClassInfoStore();
         PropertyMetadataStore metadataStore = new DefaultPropertyMetadataStore(ImmutableList.of(
             new ClasspathPropertyAnnotationHandler(), new CompileClasspathPropertyAnnotationHandler()
         ));
-        Queue<ClassNode> queue = new ArrayDeque<ClassNode>();
-        queue.add(new ClassNode(null, beanClass));
-        boolean cacheable = taskClassInfoStore.getTaskClassInfo(Cast.<Class<? extends Task>>uncheckedCast(beanClass)).isCacheable();
+        Queue<TypeNode> queue = new ArrayDeque<TypeNode>();
+        queue.add(new TypeNode(null, TypeToken.of(topLevelBean)));
+        boolean cacheable = taskClassInfoStore.getTaskClassInfo(Cast.<Class<? extends Task>>uncheckedCast(topLevelBean)).isCacheable();
 
         while (!queue.isEmpty()) {
-            ClassNode node = queue.remove();
-            validateTaskClass(beanClass, cacheable, problems, queue, node, metadataStore.getTypeMetadata(node.getBeanClass()));
+            TypeNode node = queue.remove();
+            TypeToken<?> beanType = node.getBeanType();
+            Class<?> beanClass = beanType.getRawType();
+            TypeMetadata typeMetadata = metadataStore.getTypeMetadata(beanClass);
+            if (Iterable.class.isAssignableFrom(beanClass) && !typeMetadata.isAnnotated()) {
+                @SuppressWarnings("unchecked")
+                TypeToken<Iterable> typeToken = (TypeToken<Iterable>) beanType;
+                ParameterizedType type = (ParameterizedType) typeToken.getSupertype(Iterable.class).getType();
+                TypeToken<?> nestedType = TypeToken.of(type.getActualTypeArguments()[0]);
+                queue.add(new TypeNode(node.parentPropertyName + "$1", nestedType));
```

Comment:
When doing static analysis, I think we should use the name of the type or something instead of `$1`. That would lead to better error messages.

Reply:
I added the simple name of the type.
This results now in error messages like `beans<List>.annotatedProperty` and the like.
I was thinking maybe we could add the type parameter of the `List`, too. Or maybe even only add the type parameter.
Then we would have something like `beans|NestedBean.annotatedProperty` or `beans|Set|NestedBean.annotatedProperty`.
WDYT?